Add caching, network handler improvements, and performance improvements

Opalinium opened the pull request:
Proposed changes reduce API calls by caching a players timezone for a determinate period of time (1 day) and similarly, local-time placeholders are much more consistent with a retryDelay being ran in the event that a players timezone is unable to be fetched due to API rate-limits. 

This should functionally mitigate a common occurrence of the expansion being "unable to fetch a players timezone" due to excessive API calls resulting in a rate-limit. The expansion has also been made to be completely thread-safe

Opalinium committed:
Network & API handler improvements

Improved Network connection handling via a connection read timeout, with exception handling
Improved API Response Handler - BufferedReader now reads the entire response string and checks whether it is null before assigning it to the 'timezone' variable

Opalinium committed:
Implement retryDelay to DateManager constructor & caching of players timezone via UUID

Opalinium committed:
Implement BukkitRunnable task for retryDelay

When the API call fails, added a new BukkitRunnable that will be scheduled to run after the retryDelay. It will retry adding the timezone to the map after the delay instead of adding it immediately. This way, we prevent overloading the API

Opalinium committed:
rUnNaBLE bAd

getTimeZone now uses a new thread instead of a BukkitRunnable to handle the API request

Opalinium committed:
Append license info

Opalinium committed:
Code fixes

Integrated 'ConcurrentHashMap' in place of 'Hashmap' to ensure the expansion is thread-safe
retryDelay was also migrated to a new 'getTimeZoneFromAPI' method, and appropriately initialized

Opalinium committed:
Oops

Opalinium committed:
Oops Pt. 2

I really need to sleep...

Opalinium committed:
Enhance DateManager and LocalTimeExpansion classes

Implement cache expiration in DateManager
Optimize cache structure in DateManager
Return CompletableFuture in DateManager's getTimeZone
Adapt LocalTimeExpansion to work with updated DateManager

Opalinium committed:
Refactor DateManager: separate cache expiry, data removal, and improve efficiency

Remove cache expiry checks from getTimeZone method
Remove data removal from onLeave event handler
Create a new scheduled task removeExpiredEntries to handle cache expiry and player data removal
Update isCacheExpired method to remove players from the timezones map when cache expires
Add a check in removeExpiredEntries to skip non-UUID keys in the cache
Remove unnecessary onLeave event handler and Listener implementation

Opalinium committed:
gitignore

iGabyTM commented on src/main/java/net/aboodyy/localtime/DateManager.java:
```diff
@@ -20,28 +20,37 @@
-
-    DateManager() {
-        timezones = new HashMap<>();
+    private final Map<String, String> cache;
```
Is not possible to use guava's cache instead of making your own system?

Opalinium replied:
Thank you for pointing that out. Previously I was using my own system to avoid additional dependencies, and have more granular control over the cache, however in this context that really isn't necessary. Latest commits now reflect the appropriate changes

iGabyTM commented on src/main/java/net/aboodyy/localtime/DateManager.java:
```diff
@@ -53,55 +62,124 @@ public String getDate(String format, String timezone) {
+
+
+        futureTimezone.exceptionally(ex -> {
+            Bukkit.getLogger().warning("[LocalTime] Exception while getting timezone for player " + player.getName() + ": " + ex.getMessage());
```
Use PAPI's logger to avoid confusion and keep the id of the expansion as prefix

Opalinium replied:
Corrected - previous usage of Bukkit's logger was for parity with the original source code

Opalinium committed:
Replace ConcurrentHashmap with Guava's Cache

Opalinium committed:
Use PAPI's logger instead of Bukkit's

Opalinium committed:
Fixes & Improvements

`Shutdown` method to shut down the executor service on `clear()`
Fixed `getTimeZone` method to be non-blocking
Removed premature return of `CompletableFuture`

iGabyTM commented on src/main/java/net/aboodyy/localtime/DateManager.java:
```diff
@@ -35,22 +37,23 @@
         this.timezones = new ConcurrentHashMap<>();
-        this.cache = new ConcurrentHashMap<>();
+        this.cache = CacheBuilder.newBuilder()
+                .expireAfterWrite(cacheExpirationMinutes, TimeUnit.MINUTES)
```
Looks good. This can be changed to `expireAfterWrite(1, TimeUnit.DAYS)`

Opalinium committed:
Consolidate expireAfterWrite duration

untuned commented:
Could support be added for using an API key? People on shared hosts may not be able to use this regardless if other people on the same machine are making API calls. See: #3

TomLewis commented:
I too am looking for a fix to use another IP checking URL & API to the config
